File: shared/auth/rbac.ts
import { ControlledError, Feature, PermissionAction, RequestContext, Role } from "../types/core";

type RoleAccess = Partial<Record<Feature, PermissionAction[]>>;

export const roleAccessMatrix: Record<Role, RoleAccess> = {
  super_admin: {
    platform: ["manage"],
    schools: ["view", "create", "update", "delete", "manage"],
    audit_logs: ["view"],
    reports: ["view"]
  },
  admin: {
    users: ["view", "create", "update", "delete", "manage"],
    settings: ["view", "create", "update", "delete", "manage"],
    students: ["view", "create", "update", "delete", "manage"],
    teachers: ["view", "create", "update", "delete", "manage"],
    subjects: ["view", "create", "update", "delete", "manage"],
    classes: ["view", "create", "update", "delete", "manage"],
    attendance: ["view", "create", "update", "delete", "manage"],
    homework: ["view", "create", "update", "delete", "manage"],
    exams: ["view", "create", "update", "delete", "manage"],
    results: ["view", "create", "update", "delete", "manage"],
    fees: ["view", "create", "update", "delete", "manage"],
    reports: ["view"],
    notifications: ["view", "create", "update"],
    audit_logs: ["view"],
    announcements: ["view", "create", "update", "delete", "manage"],
    timetable: ["view", "create", "update", "delete", "manage"],
    behavior: ["view", "create", "update", "delete", "manage"],
    leave: ["view", "create", "update", "delete", "manage"],
    events: ["view", "create", "update", "delete", "manage"]
  },
  teacher: {
    settings: ["view"],
    teachers: ["view"],
    subjects: ["view"],
    classes: ["view"],
    attendance: ["view", "create", "update"],
    homework: ["view", "create", "update"],
    exams: ["view", "create", "update"],
    results: ["view", "create", "update"],
    reports: ["view"],
    notifications: ["view"],
    announcements: ["view"],
    timetable: ["view"],
    behavior: ["view", "create", "update"],
    leave: ["view", "create"],
    events: ["view"]
  },
  parent: {
    settings: ["view"],
    subjects: ["view"],
    classes: ["view"],
    attendance: ["view"],
    homework: ["view"],
    exams: ["view"],
    results: ["view"],
    fees: ["view"],
    reports: ["view"],
    notifications: ["view"],
    announcements: ["view"],
    timetable: ["view"],
    behavior: ["view"],
    events: ["view"]
  }
};

export function canAccess(role: Role, feature: Feature, action: PermissionAction): boolean {
  const allowed = roleAccessMatrix[role][feature] ?? [];
  return allowed.includes(action) || allowed.includes("manage");
}

export function assertPermission(
  ctx: RequestContext,
  feature: Feature,
  action: PermissionAction
): void {
  const explicit = ctx.permissions.includes(`${feature}:${action}`) || ctx.permissions.includes(`${feature}:manage`);
  if (!explicit && !canAccess(ctx.role, feature, action)) {
    throw new ControlledError("FORBIDDEN", "You do not have permission to perform this action.", 403, {
      feature,
      action
    });
  }
}

File: backend-go/internal/auth/rbac.go
package auth

import "github.com/eduplexo/backend-go/internal/api"

// Action is one of the five permission verbs from
// old-app/shared/types/core.ts.
type Action string

const (
	ActionView   Action = "view"
	ActionCreate Action = "create"
	ActionUpdate Action = "update"
	ActionDelete Action = "delete"
	ActionManage Action = "manage"
)

// Feature mirrors the `Feature` union from
// old-app/shared/types/core.ts.
type Feature string

// roleAccessMatrix is ported verbatim from
// old-app/shared/auth/rbac.ts. Each role maps to a set of features and the
// actions allowed on each feature. `manage` is treated as a superset of any
// other action by `CanAccess`.
var roleAccessMatrix = map[string]map[Feature][]Action{
	"super_admin": {
		"platform":   {ActionManage},
		"schools":    {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"audit_logs": {ActionView},
		"reports":    {ActionView},
	},
	"admin": {
		"users":         {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"settings":      {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"students":      {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"teachers":      {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"subjects":      {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"classes":       {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"attendance":    {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"homework":      {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"exams":         {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"results":       {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"fees":          {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"reports":       {ActionView},
		"notifications": {ActionView, ActionCreate, ActionUpdate},
		"audit_logs":    {ActionView},
		"announcements": {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"timetable":     {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"behavior":      {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"leave":         {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
		"events":        {ActionView, ActionCreate, ActionUpdate, ActionDelete, ActionManage},
	},
	"teacher": {
		"settings":      {ActionView},
		"students":      {ActionView},
		"teachers":      {ActionView},
		"subjects":      {ActionView},
		"classes":       {ActionView},
		"attendance":    {ActionView, ActionCreate, ActionUpdate},
		"homework":      {ActionView, ActionCreate, ActionUpdate},
		"exams":         {ActionView, ActionCreate, ActionUpdate},
		"results":       {ActionView, ActionCreate, ActionUpdate},
		"fees":          {ActionView},
		"reports":       {ActionView},
		"notifications": {ActionView},
		"announcements": {ActionView},
		"timetable":     {ActionView},
		"behavior":      {ActionView, ActionCreate, ActionUpdate},
		"leave":         {ActionView, ActionCreate},
		"events":        {ActionView},
	},
	"parent": {
		"settings":      {ActionView},
		"subjects":      {ActionView},
		"classes":       {ActionView},
		"attendance":    {ActionView},
		"homework":      {ActionView},
		"exams":         {ActionView},
		"results":       {ActionView},
		"fees":          {ActionView},
		"reports":       {ActionView},
		"notifications": {ActionView},
		"announcements": {ActionView},
		"timetable":     {ActionView},
		"behavior":      {ActionView},
		"events":        {ActionView},
	},
	"student": {
		"settings":      {ActionView},
		"subjects":      {ActionView},
		"classes":       {ActionView},
		"attendance":    {ActionView},
		"homework":      {ActionView},
		"exams":         {ActionView},
		"results":       {ActionView},
		"fees":          {ActionView},
		"reports":       {ActionView},
		"notifications": {ActionView},
		"announcements": {ActionView},
		"timetable":     {ActionView},
		"events":        {ActionView},
		"leave":         {ActionView, ActionCreate},
	},
}

// CanAccess returns true when `role` may perform `action` on `feature`.
// `manage` implicitly grants every action.
func CanAccess(role string, feature Feature, action Action) bool {
	allowed, ok := roleAccessMatrix[role][feature]
	if !ok {
		return false
	}
	for _, a := range allowed {
		if a == action || a == ActionManage {
			return true
		}
	}
	return false
}

// AssertPermission mirrors the Node `assertPermission()` helper. Returns a
// ControlledError(FORBIDDEN, 403) when the requesting user is missing the
// permission. Explicit per-user permissions on the JWT (`feature:action`)
// override the role default.
func AssertPermission(ctx *api.RequestContext, feature Feature, action Action) error {
	wantFA := string(feature) + ":" + string(action)
	wantManage := string(feature) + ":manage"
	for _, p := range ctx.Permissions {
		if p == wantFA || p == wantManage || p == "*" {
			return nil
		}
	}
	if CanAccess(ctx.Role, feature, action) {
		return nil
	}
	return api.NewControlledError(
		"FORBIDDEN",
		"You do not have permission to perform this action.",
		403,
		map[string]string{"feature": string(feature), "action": string(action)},
	)
}
